README.md
# ZX Spectrum Next

Xilinx ISE 14.7 is required to build the zx spectrum next core.

A 16MB flash chip connected to the fpga is used to store core bitstreams.  The chip is divided into 32 slots of 512KB each, each of which can store a bitstream.  The first two slots, slot 0 and slot 1, are reserved for the zx next.  Slot 0 contains the anti-brick core, a core intended not to be changed as it is used to rescue the machine if something goes wrong.  Slot 1 contains the zx next core.

The fpga will configure itself using the anti-brick core in slot 0 at power up.  Shortly after, the fpga will reconfigure using the zx next core in slot 1 and the spectrum next will boot per normal after customizing itself using information stored in the file **machines/next/config.ini** on the sd card.

The vhdl source code can be found rooted in **src/**
* **zxnext_pins_issue2.ucf** defines the external pinout of the fpga and the build constraints.
* **zxnext_top_issue2.vhd** contains connections to the external pins and takes care of housekeeping not related to the zx next machine itself.
* **zxnext.vhd** is the implementation of the zx spectrum next. The source code for instantiated modules can be found in surrounding subdirectories.

## BUILDING TBBLUE.TBU

TBBLUE.TBU is a file that contains a header and a zx next core bitstream.  The bitstream stored in this file is copied to slot 1 in the flash chip.  The header contains a crc check and a core version number.

To generate TBBLUE.TBU, first build the zx next core.

1. If you have made changes to the vhdl, make sure the core version number is set correctly at the top of **src/zxnext_top_issue2.vhd** at about line 50.
2. Navigate to **synth-zxnext-issue2/**
3. Double click on **zxnext-issue2.xise** to have ISE automatically load the project.  If this file type is not associated with ISE on your machine, start ISE as you normally would and load this project file from the menu.
4. Right click on "Generate Programming File" as shown in the diagram below and choose "Re-Run All".  You may have to choose the "Implementation View" in the top left and the "Design" tab on the bottom left to see the same panels as shown in the picture.

![Image: Xilinx ISE Screenshot](zxnext-ise-build.png "Xilinx ISE Screenshot")

5. After some time, the build will complete with the bitstream generated into **synth-zxnext-issue2/work/zxnext_top_issue2.bit**
6. Navigate to **tbumaker/**
7. Open a command prompt in that directory.  (On windows this may be "FILE->Open Command Prompt")
8. Run "maketbu"
9. Where it asks for a version number, enter the core id as set in step 1 (eg 30100 = 3.01.00)
10. The file "TBBLUE.TBU" will be generated which can then be copied to the root directory on the sd card, replacing the TBBLUE.TBU file already present.

The TBBLUE.TBU file is only used when updating the machine.  The fpga configures itself using the bitstream stored in slot 1 of the flash.  The updater reads the TBBLUE.TBU file, verifies the crc is correct, and then writes it to slot 1.  So the final step to update your machine is to run the updater.

If you would like to perform an anti-brick update, follow the instructions listed in the manual.  An anti-brick update can be used if the zx next core is damaged or bugged such that the zx next core is not operating correctly.  If the zx next core is not operating correctly, then it is best to use the anti-brick core to perform the update to make sure the flash is not incorrectly written.  During an anti-brick update, the machine may only generate a VGA picture.

To perform a regular update using the zx next core, hold "U" while powering up or while doing a hard reset.  The updater screen will appear, let go of "U" and answer yes to perform the update.  You must not interrupt the update process; interrupting the update could damage the contents of slot 1 so that the machine cannot boot up.  Then an anti-brick update would have to be performed to fix slot 1.  When the update is finished, completely power down the machine by disconnecting the power supply and disconnecting hdmi if you are using that.  Wait about 10 seconds, plug in the hdmi and then power up the machine.  If all goes well, the update will be completed.

## BUILDING AN MCS FILE

An mcs file can be used to program the flash chip directly using a jtag interface and cable.  This method does not require the zx next to be on or even functioning.

The mcs file built will hold the contents of slot 0 (the anti-brick core) and slot 1 (the zx next core).

After successfully building the zx next core (see the last section), a file **synth-zxnext-issue2/work/zxnext_top_issue2.bit** is generated.

1. Copy **synth-zxnext-issue2/work/zxnext_top_issue2.bit** to **mcsmaker/**.  There should now be two bit files **next_ab_issue2_top.bit** and **zxnext_top_issue2.bit**.
2. Open a command prompt in **mcsmaker/**
3. Run "mcs_generator_issue2".
4. The mcs file "zxnext.mcs" will be produced.

## SPECIFICATIONS

The ZX Spectrum Next core is not completed yet but the aim is to finish for the second kickstarter.  Nevertheless the majority of the planned functionality is already implemented.

A brief summary of the current machine specifications:

* **CPU** : Z80N (z80 compatible with some additional instructions) operable at software selectable speeds of 3.5 MHz, 7 MHz, 14 MHz or 28 MHz with wait states.
* **COPPER** : A co-processor running independently of the cpu executes simple instructions that can modify the nextreg state.  For example, it can change palettes, alter the display mode or play stereo music.  The copper is synchronized with the display generation so it is able to make these changes at precise locations in the display.
* **DMA** : The ZXN DMA, compatible with a subset of the Z80 DMA chip, is able to perform transfers between memory and/or io using short two cycle reads and writes.  In burst mode, the zxn dma can send bytes at programmable rates allowing it to play sampled music while returning control to the cpu between transfers.  At this time the zxn dma and the cpu share the bus and the dma operates at the currently set cpu speed.
*  **RAM MEMORY** : 768K of RAM in the unexpanded machine or 1792K of RAM in the expanded machine.   This memory is available in 16K banks as in the original 128K Spectrums and can be mapped as usual using the standard ports 0x7ffd and 0x1ffd.   An additional port 0xdffd adds bits to port 0x7ffd to reach all memory banks.  The native bankswitching scheme in the zx next is called MMU.  This scheme divides the same memory into 8K pages and allows any page to be mapped into any 8K slot of the Z80's 64K address space.
*  **ROM MEMORY** : 64K of ROM is reserved for ROMs 0-3 as in the Spectrum +3.  Also available is 32K of Alt ROM that can replace the normal ROMs; this ROM is user programmable.
* **GRAPHICS** : The display is composed of layers with programmable priority.  Layers are listed below.
	* ULA : Compatible timing, contention and floating bus behaviour with the 48k, 128k, +3 and Pentagon.  Supports hardware pixel scrolling in the X and Y directions and these resolutions:
		* 256x192 pixel 32x24 attributes in bank 5 (48k / 128k)
		* 256x192 pixel 32x24 attributes in bank 7 (128k second display)
		* 256x192 pixel 32x24 attributes at 0x6000 in bank 5 (timex second display)
		* 256x192 pixel 32x192 attributes at 0x4000 and 0x6000 in bank 5 (timex hi-colour)
		* 512x192 pixel monochrome at 0x4000 and 0x6000 in bank 5 (timex hi-res)
	* LoRes : Occupies the same layer as the ULA with LoRes replacing the ULA where it is enabled.  Supports hardware pixel scrolling in the X and Y directions.  Two resolutions are available:
		* 128x96 4-bit colour per pixel at either 0x4000 or 0x6000 in bank 5 (Radastan mode originating on the ZX UNO)
		* 128x96 8-bit colour per pixel occupying 0x4000 and 0x6000 in bank 5
	* Layer 2 : A pixel mapped display without colour clash.  Supports hardware pixel scrolling in the X and Y directions.  Can be mapped to any location in memory starting at a 16K boundary.  Available resolutions are:
		* 256x192 8-bit colour per pixel
		* 320x256 8-bit colour per pixel
		* 640x256 4-bit colour per pixel
	* Sprites : Up to 128 hardware sprites of size 16x16 pixels with either 8-bits or 4-bits of colour per pixel.  A minimum of 100 sprites per line can be display at this size.  Sprites can be scaled 1x 2x 4x 8x, rotated, mirrored and linked together.
	* Tilemap : A hardware character display coming in two resolutions (80x32 = 640x256 pixels, 40x32 = 320x256).  Supports hardware pixel scrolling in the X and Y directions.  The character map and glyphs are stored at programmable locations in bank 5.  Individual characters can be independently rotated and mirrored.  Each glyph is 8x8 pixels in size with 4-bits of colour per pixel.  Another mode eliminates rotation and mirroring in favour of 8x8 pixel glyphs defined as monochrome UDGs but with more colour information stored in the character map.

	Some layer priority modes allow layer 2 to be highlighted or darkened by the ULA.  Another setting allows the tilemap and ULA to stencil each other.
* **SOUND** : Stereo sound is played through HDMI, 3.5mm audio jack or optional internal speaker.  Sound sources:
	* Beeper : Beeps and tape sound.
	* 3 x AY 8910 : Arranged to be compatible with the dual arrangement turbosound.  AY instances can be programmed as mono or either ABC or ACB stereo mode.
	* 4 x 8-bit DACs : Two DACs are assigned to the left channel and two are assigned to the right channel.  Common 8-bit dac peripherals in the spectrum community such as specdrum and soundrive are mapped to these dacs.
	* Raspberry PI I2S : Audio generated by an optional Pi accelerator can be mixed into the internal next audio stream or can be mapped to ear for tape loading.




changelog.md
|Release|Author          |Description                                       |  
|-------|----------------|--------------------------------------------------|  
|3.00.XX|Allen Albright  |Refactor (Many Changes)  |  
|3.01.00|Allen Albright  |First Public Release (Day Zero Update)            |
|3.01.01|Allen Albright  |<b>ULA</b> : Interrupt time in 48K and 128K video frame was off by one cycle<br/><b>ULA+</b> : Byte read from port 0xff3b incorrectly indicated ula+ was not present<br/><b>AUDIO</b> : Attempt to double volume on 3.5mm jack and internal speaker<br/><b>AY</b> : Implement differences in how AY and YM respond to register reads<br/><b>PORT DECODE</b> : Bit 31 of nextreg 0x89 and 0x85 determines whether the port decodes are soft or hard reset<br/><b>SCANLINES</b> : Change weights to 0, 12.5, 25, 50<br/><b>COMPATIBILITY</b> : Port decoding style now determined by current video timing rather than static machine type; active ports no longer limited by machine type<br/><b>STICKS</b> : Sinclair 1 and Sinclair 2 joystick types were reversed in the hardware<br/><b>HDMI</b> : Audio stream made signed instead of unsigned<br/><b>NMI</b> : Function key activation now cancels nmi generation so that minimum press time of one second is no longer needed<br/><b>MEMORY</b> : Nextreg 0x8e added to more quickly perform traditional bankswitching normally done through ports 0x1ffd, 0x7ffd, 0xdffd<br/><b>SLU 6 & 7</b> : Blending updated to be more useful<br/>* Nextreg 0x68 bits 6:5 control which of ula or tilemap layer blends with layer 2<br/>* SLU 6 = (U or T)S(T or U)(B+L) colours clamped to [0,7]<br/>* SLU 7 = (U or T)S(T or U)(B+L-5) colours clamped to [0,7]<br/>In both cases U or T only appear independently if they are not involved in blending     |  
|3.01.02|Allen Albright  |<b>DMA</b> : Ports 0x6b and 0x0b (new) both address the dma. When port 0x0b is read or written the dma is placed in z80-dma compatibility mode. When port 0x6b is read or written the dma is placed in its native zxn-dma mode.  Nextreg 0x06 bit 6 which used to select dma mode has been removed.<br/><b>PORT DECODE</b> : Nextreg 0x85 and 0x89 have bit 25 added to enable dma port 0x0b<br/><b>EXPANSION BUS</b> : Nextreg 0x8a bit 4 allows propagation of port 0xff io cycles to the expansion bus.<br/><b>AUDIO</b> : Nextreg 0x06 bit 6 chosen to select a new classic audio mode where beep and tape noises are directed to the internal speaker exclusively and other audio is directed to hdmi and the 3.5mm jack.<br/><b>AUDIO</b> : Another adjustment of the internal speaker volume and a different treatment for the classic audio mode that allows beep noises to be louder.  |
|3.01.03|Allen Albright  |<b>AY</b> : Fix bug where the envelope period counter was not reset when register 13 was written.<br/><b>AUDIO</b> : Another adjustment of internal speaker volume with total volume increased at the expense of increased likelihood of clipping and classic mode audio made a little more quiet.<br/><b>EXPANSION BUS</b> : (trial) Introduce a ROMCS rom replacement feature enabled by bits 6 and 2 in nextreg 0x80. External devices can continue to assert /romcs but the next can replace the 16k external rom with divmmc banks 14 and 15 to allow custom code to be written to support external devices more seemlessly.<br/><b>EXPANSION BUS</b> : Nextreg 0x8a defaults to all port propagation disabled on hard reset. Port 0xfe was propagated by default prior to this change.<br/><b>MEMORY</b> : Nextreg 0x8e no longer changes mmu 6/7 when the ram bank is not changed.  |
|3.01.04|Allen Albright  |<b>DMA</b> : The 16-bit byte count now read in the correct endian order through the read register.<br/><b>MULTIFACE</b> : Fix numerous errors in the implementation of the mf128 and mf48.<br/><b>STICKS</b> : Correct md pad state machine that was reading some input at the wrong time.<br/><b>STICKS</b> : The joystick connectors can now be configured as an i/o device with 1 output and up to 12 inputs. Three modes are supported including bit-bang where the state of the output bit is controlled by the program, clock where the output bit is driven by either a slow or a fast clock, and uart where one of the next's two internal uarts can be redirected to the joystick connector. Details can be found in documentation for port 0x37 and nextreg 0x05.<br/><b>STICKS</b> : Button A on md pads now generates an UP direction for non- md pad joystick modes.<br/><b>COMPATIBILITY</b> : Pentagon 512K support is added that is activated when the video timing is set to Pentagon and when bit 7 of port 0xdffd is set to 1.  In this mode, port 0xdffd is ignored in banking and instead bits 7:6 of port 0x7ffd extend bits 2:0 of the same port to select a 16K bank.<br/><b>KEYBOARD</b> : It is now possible to read the state of the new composite keys directly instead of indirectly through the key membrane. Details under nextreg 0x68, 0xb0, 0xb1.  |

utils/romgen.cpp

#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <math.h>
#include <iostream>

using namespace std;

#define MAX_ROM_SIZE 0x4000

static unsigned int mylog2 (unsigned int val) {
	unsigned int ret = -1;
	while (val != 0) {
		val >>= 1;
		ret++;
	}
	return ret;
}

int main(int argc, char* argv[])
{
	cerr << "romgen by MikeJ version 3.00\n";
	// read file

	if(argc < 3)
	{
		cerr << "\nUsage: romgen <input file> <entity name> \n";
		cerr << "                  {format} {registered} {enable}\n";
		cerr << "\n";
//		cerr << "Now uses bit_vector generics for compatibilty with Xilinx UniSim library\n";
//		cerr << "\n";
		cerr << "for the format paramater use :\n";
		cerr << "  a  - rtl model rom array (standard)\n";
		cerr << "  c  - rtl model case statement \n";
		cerr << "  b  - Xilinx block ram4  (registered output always)\n";
		cerr << "  l  - Xilinx block ram16 (registered output always)\n";
//		cerr << "  d  - Xilinx distributed ram [not supported yet]\n";
		cerr << "  m  - Xilinx mem file\n";
		cerr << "\n";
		cerr << "for the registered paramater (optional) use :\n";
		cerr << "  r  - registered output (combinatorial otherwise) \n";
		cerr << "\n";
		cerr << "for the enable paramater (optional) use :\n";
		cerr << "  e  - clock enable generated \n";
		cerr << "\n";
		cerr << "note, generated roms are always 8 bits wide\n";
		cerr << "      max 12 address bits for block ram4s\n";
		cerr << "      max 14 address bits for block ram16s\n";
		cerr << "for example romgen fred.bin fred_rom c r\n\n";
		return -1;
	}

	string buffer;
	FILE *fin;

	fin = fopen(argv[1], "rb");
	if (fin == NULL) {
	  cerr << "ERROR : Could not open input file " << argv[1] <<"\n";
	  return -1;
	}

	fseek(fin, 0, SEEK_END );
	int FileSize = ftell( fin );
	rewind( fin );
	char rom_type = 'a';
	char option_1 = 0;
	char option_2 = 0;
	if (argc > 3) sscanf(argv[3],"%c",&rom_type);
	if (argc > 4) sscanf(argv[4],"%c",&option_1);
	if (argc > 5) sscanf(argv[5],"%c",&option_2);

	bool format_case = false;
	bool format_array = false;
	bool format_block = false;
	bool format_dist = false;
	bool format_clock = false;
	bool format_ram16 = false;
	bool format_ena = false;
	bool format_mem = false;

	cerr << "INFO : creating entity : " << argv[2] << "\n";

	if (option_1 != 0) {
	  if ((option_1 == 'r') || (option_1 == 'R'))
		format_clock = true;
	  else if ((option_1 == 'e') || (option_1 == 'E'))
		format_ena = true;
	  else {
		cerr << "ERROR : output option not supported\n";
		return -1;
	  }
	}

	// lazy ...
	if (option_2 != 0) {
	  if ((option_2 == 'r') || (option_2 == 'R'))
		format_clock = true;
	  else if ((option_2 == 'e') || (option_2 == 'E'))
		format_ena = true;
	  else {
		cerr << "ERROR : output option not supported\n";
		return -1;
	  }
	}

	if ((rom_type == 'c') || (rom_type == 'C')) {
		cerr << "INFO : rtl model, case statement \n"; format_case = true; }
	else if ((rom_type == 'a') || (rom_type == 'A')) {
		cerr << "INFO : rtl model, rom array \n"; format_array = true; }
	else if ((rom_type == 'b') || (rom_type == 'B')) {
		cerr << "INFO : block4 ram, registered \n"; format_block = true; format_clock = true; }
	else if ((rom_type == 'l') || (rom_type == 'L')) {
		cerr << "INFO : block16 ram, registered \n"; format_block = true; format_clock = true; format_ram16 = true; }
	else if ((rom_type == 'm') || (rom_type == 'M')) {
		cerr << "INFO : mem file \n"; format_mem = true; format_clock = true; format_ram16 = true;}
	//else if ((rom_type == 'd') || (rom_type == 'D')) {
	//  cerr << "INFO : distributed ram, combinatorial; \n"; format_dist = true; }
	else {
	  cerr << "ERROR : format not supported\n";
	  return -1;
	}
	if (format_clock == true)
		cerr << "INFO : registered output\n\n";
	else
		cerr << "INFO : combinatorial output\n\n";


	// calc number of address bits required
	int addr_bits;
	int max_addr_bits = 16;
	int rom_inits = 16;

	if (format_block == true || format_mem == true) {
	  if (format_ram16 == true) {
		max_addr_bits = 14;
		rom_inits = 64;
	  }
	  else
		max_addr_bits = 12;
	}

	addr_bits = mylog2(FileSize);
/*
	sscanf(argv[3],"%d",&addr_bits);
*/
	if (addr_bits < 1 || addr_bits > max_addr_bits) {
	  cerr << "ERROR : illegal rom size, number of address bits must be between 1 and " << max_addr_bits << "\n";
	  return -1;
	}

	// ram b16s
	// for  14 bits use ram_b16_s1 x data_width
	// for  13 bits use ram_b16_s2 x data_width/2
	// for  12 bits use ram_b16_s4 x data_width/4
	// for<=11 bits use ram_b16_s8 x data_width/8

	// ram b4s
	// for  12 bits use ram_b4_s1 x data_width
	// for  11 bits use ram_b4_s2 x data_width/2
	// for  10 bits use ram_b4_s4 x data_width/4
	// for <=9 bits use ram_b4_s8 x data_width/8
	int rom_size = (int) pow(2,addr_bits);

	int number_of_block_rams = 1;
	int block_ram_width = 8;
	int block_ram_pwidth = 0;
	int block_ram_abits = 9;

	if (format_ram16 == true) {
	  block_ram_abits = 11; // default
	  block_ram_pwidth = 1;
	  // ram16s
	  switch (addr_bits) {
	   case 14 : number_of_block_rams = 8; block_ram_width = 1; block_ram_pwidth = 0; block_ram_abits = 14; break;
	   case 13 : number_of_block_rams = 4; block_ram_width = 2; block_ram_pwidth = 0; block_ram_abits = 13; break;
	   case 12 : number_of_block_rams = 2; block_ram_width = 4; block_ram_pwidth = 0; block_ram_abits = 12; break;
	   default : ;
	  }
	}
	else {
	  // ram4s
	  switch (addr_bits) {
	   case 12 : number_of_block_rams = 8; block_ram_width = 1; block_ram_abits = 12; break;
	   case 11 : number_of_block_rams = 4; block_ram_width = 2; block_ram_abits = 11; break;
	   case 10 : number_of_block_rams = 2; block_ram_width = 4; block_ram_abits = 10; break;
	   default : ;
	  }
	}

	//printf("block ram w : %d ",block_ram_width);
	//printf("block ram n : %d ",number_of_block_rams);


	// process
	int mem[MAX_ROM_SIZE];
	string line;

	int i,j,k;


	int addr = 0;
	int last_nz_addr = 0;
	int offset = 0;
	int mask = 0;
	unsigned int data = 0;

	// clear mem
	for (i = 0; i < MAX_ROM_SIZE; i++) mem[i] = 0;

	// process file
	data = getc(fin);
	while (!feof(fin) && (addr < rom_size)) {
	  if (addr >= MAX_ROM_SIZE) {
		cerr << "ERROR : file too large\n";
		return -1;
	  }

	  mem[addr] = data;
	  // debug
	  //if (addr % 16 == 0) printf("%04x : ",addr);
	  //printf("%02x  ",data);
	  //if (addr % 16 == 15) printf("\n");
	  // end debug
	  if (data) {
			// remember the last address that has non zero data
	  	last_nz_addr = addr;
	  }
	  addr ++;
	  data = getc(fin);
	}
	fclose(fin);


	if (format_mem == true) {


	  for (k = 0; k < number_of_block_rams; k ++){
		//printf("  rom%d : if true generate\n",k);

		for (j = 0; j < rom_inits; j++) {
		  //printf("    attribute INIT_%02X of inst : label is \042",j);
		  switch (block_ram_width) {

		  case 1 : // width 1
		  mask = 0x1 << (k);
		  //for (i = 0; i < 256; i+=8) {
		  for (i = 248; i >= 0; i-=8) {
			data  = ((mem[(j*256) + (255 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (254 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (253 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (252 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (251 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (250 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (249 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (248 - i)] & mask) >> k);
			printf("%02X ",data);
		  }
		  break;

		  case 2 : // width 2
		  //printf("case 2\n");
		  mask = 0x3 << (k * 2);
		  //for (i = 0; i < 128; i+=4) {
		  for (i = 124; i >= 0; i-=4) {
			data  = ((mem[(j*128) + (127 - i)] & mask) >> k * 2);
			data <<= 2;
			data += ((mem[(j*128) + (126 - i)] & mask) >> k * 2);
			data <<= 2;
			data += ((mem[(j*128) + (125 - i)] & mask) >> k * 2);
			data <<= 2;
			data += ((mem[(j*128) + (124 - i)] & mask) >> k * 2);
			printf("%02X ",data);
		  }
		  break;

		  case 4 : // width 4
		  printf("case 4\n");
		  mask = 0xF << (k * 4);
		  for (i = 0; i < 64; i+=2) {
			data  = ((mem[(j*64) + (63 - i)] & mask) >> k * 4);
			data <<= 4;
			data += ((mem[(j*64) + (62 - i)] & mask) >> k * 4);

			printf("%02X ",data);
		  }
		  break;


		  case 8 : // width 8
		  //for (i = 0; i < 32; i++) {
		  //printf("case 8\n");
		  for (i = 31; i >= 0; i--) {
			data = ((mem[(j*32) + (31 - i)]));
			printf("%02X ",data);
		  }
		  break;
		  } // end switch

		  //printf("\042;\n");
		}

	  // for (i = 0; i <= last_nz_addr; i ++ ) {
		// printf("%02X",i,mem[i]);
	  }
	  return 0;
	} // end case

	printf("-- generated with romgen v3.0 by MikeJ\n");
	printf("library ieee;\n");
	printf("use ieee.std_logic_1164.all;\n");
	printf("use ieee.std_logic_unsigned.all;\n");
	printf("use ieee.numeric_std.all;\n");
	printf("\n");
//	printf("library UNISIM;\n");
//	printf("  use UNISIM.Vcomponents.all;\n");
//	printf("\n");
	printf("entity %s is\n",argv[2]);
	printf("\tport (\n");
	if (format_clock == true) printf("\t\tclk\t\t: in    std_logic;\n");
	if (format_ena == true)   printf("\t\tena\t\t: in    std_logic;\n");
	printf("\t\taddr\t\t: in    std_logic_vector(%d downto 0);\n",addr_bits - 1);
	printf("\t\tdata\t\t: out   std_logic_vector(7 downto 0)\n");
	printf("\t);\n");
	printf("end;\n");
	printf("\n");
	printf("architecture rtl of %s is\n",argv[2]);
	printf("\n");

	// if blockram
	//{{{
	if (format_block == true) {
	  printf("  function romgen_str2bv (str : string) return bit_vector is\n");
	  printf("    variable result : bit_vector (str'length*4-1 downto 0);\n");
	  printf("  begin\n");
	  printf("    for i in 0 to str'length-1 loop\n");
	  printf("      case str(str'high-i) is\n");
	  for (i = 0; i<16; i++)
		printf("        when '%01X'       => result(i*4+3 downto i*4) := x\042%01X\042;\n",i,i);
	  printf("        when others    => null;\n");
	  printf("      end case;\n");
	  printf("    end loop;\n");
	  printf("    return result;\n");
	  printf("  end romgen_str2bv;\n");
	  printf("\n");

	  // xilinx block ram component
	  if (block_ram_pwidth != 0) {
		for (i = 0; i< 8; i++)
		  printf("  attribute INITP_%02X : string;\n",i);
		printf("\n");
	  }

	  for (i = 0; i< rom_inits; i++)
		printf("  attribute INIT_%02X : string;\n",i);
	  printf("\n");

	  if (format_ram16 == true)
		printf("  component RAMB16_S%d\n",block_ram_width + block_ram_pwidth);
	  else
		printf("  component RAMB4_S%d\n",block_ram_width);

	  printf("    --pragma translate_off\n");
	  printf("    generic (\n");
	  if (block_ram_pwidth != 0) {
		for (i = 0; i< 8; i++) {
		  printf("      INITP_%02X : bit_vector (255 downto 0) := x\0420000000000000000000000000000000000000000000000000000000000000000\042",i);
		  printf(";\n");
		}
		printf("\n");
	  }

	  for (i = 0; i< rom_inits; i++) {
		printf("      INIT_%02X : bit_vector (255 downto 0) := x\0420000000000000000000000000000000000000000000000000000000000000000\042",i);
		if (i< (rom_inits - 1)) printf(";");
		printf("\n");
	  }
	  printf("      );\n");
	  printf("    --pragma translate_on\n");
	  printf("    port (\n");
	  printf("      DO    : out std_logic_vector (%d downto 0);\n",block_ram_width -1);
	  if (block_ram_pwidth != 0)
		printf("      DOP   : out std_logic_vector (%d downto 0);\n",block_ram_pwidth -1);
	  printf("      ADDR  : in  std_logic_vector (%d downto 0);\n",block_ram_abits -1);
	  printf("      CLK   : in  std_logic;\n");
	  printf("      DI    : in  std_logic_vector (%d downto 0);\n",block_ram_width -1);
	  if (block_ram_pwidth != 0)
		printf("      DIP   : in  std_logic_vector (%d downto 0);\n",block_ram_pwidth -1);
	  printf("      EN    : in  std_logic;\n");
	  if (format_ram16 == true)
		printf("      SSR   : in  std_logic;\n");
	  else
		printf("      RST   : in  std_logic;\n");
	  printf("      WE    : in  std_logic \n");
	  printf("      );\n");
	  printf("  end component;\n");
	  printf("\n");
	  printf("  signal rom_addr : std_logic_vector(%d downto 0);\n",block_ram_abits - 1);
	  printf("\n");
	}
	//}}}
	//{{{
	if (format_array == true) {
//	  printf("\n");
	  printf("\ttype ROM_ARRAY is array(0 to %d) of std_logic_vector(7 downto 0);\n",rom_size - 1);
	  printf("\tconstant ROM : ROM_ARRAY := (\n");
	  for (i = 0; i < rom_size; i ++ ) {
		if (i % 8 == 0) printf("\t\t");
		printf("x\042%02X\042",mem[i]);
		if (i  < (rom_size - 1)) printf(",");
		if (i == (rom_size - 1)) printf(" ");
		if (i % 8 == 7) printf(" -- 0x%04X\n",i - 7);
	  }
	  printf("\t);\n");
	  printf("\n");
	} // end array
	//}}}
	//{{{
	if (format_case == true) {
	  printf("  signal rom_addr : std_logic_vector(11 downto 0);\n");
	  printf("\n");
	}
	//}}}

	printf("begin\n");
	printf("\n");
	//
	if ((format_block == true) || (format_case == true)) {
	  printf("  p_addr : process(ADDR)\n");
	  printf("  begin\n");
	  printf("     rom_addr <= (others => '0');\n");
	  printf("     rom_addr(%d downto 0) <= ADDR;\n",addr_bits - 1);
	  printf("  end process;\n");
	  printf("\n");
	}
	//
	//{{{
	if (format_block == true) {
	  for (k = 0; k < number_of_block_rams; k ++){
		printf("  rom%d : if true generate\n",k);

		for (j = 0; j < rom_inits; j++) {
		  printf("    attribute INIT_%02X of inst : label is \042",j);
		  switch (block_ram_width) {

		  case 1 : // width 1
		  mask = 0x1 << (k);
		  for (i = 0; i < 256; i+=8) {
			data  = ((mem[(j*256) + (255 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (254 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (253 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (252 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (251 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (250 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (249 - i)] & mask) >> k);
			data <<= 1;
			data += ((mem[(j*256) + (248 - i)] & mask) >> k);
			printf("%02X",data);
		  }
		  break;

		  case 2 : // width 2
		  mask = 0x3 << (k * 2);
		  for (i = 0; i < 128; i+=4) {
			data  = ((mem[(j*128) + (127 - i)] & mask) >> k * 2);
			data <<= 2;
			data += ((mem[(j*128) + (126 - i)] & mask) >> k * 2);
			data <<= 2;
			data += ((mem[(j*128) + (125 - i)] & mask) >> k * 2);
			data <<= 2;
			data += ((mem[(j*128) + (124 - i)] & mask) >> k * 2);
			printf("%02X",data);
		  }
		  break;

		  case 4 : // width 4
		  mask = 0xF << (k * 4);
		  for (i = 0; i < 64; i+=2) {
			data  = ((mem[(j*64) + (63 - i)] & mask) >> k * 4);
			data <<= 4;
			data += ((mem[(j*64) + (62 - i)] & mask) >> k * 4);

			printf("%02X",data);
		  }
		  break;


		  case 8 : // width 8
		  for (i = 0; i < 32; i++) {
			data = ((mem[(j*32) + (31 - i)]));
			printf("%02X",data);
		  }
		  break;
		  } // end switch

		  printf("\042;\n");
		}

		printf("  begin\n");
		if (format_ram16 == true)
		  printf("  inst : RAMB16_S%d\n",block_ram_width + block_ram_pwidth);
		else
		  printf("  inst : RAMB4_S%d\n",block_ram_width);

		printf("      --pragma translate_off\n");
		printf("      generic map (\n");

		if (block_ram_pwidth != 0) {
		  for (i = 0; i< 8; i++) {
			printf("        INITP_%02X => x\0420000000000000000000000000000000000000000000000000000000000000000\042",i);
			printf(",\n");
		  }
		  printf("\n");
		}

		for (i = 0; i< rom_inits; i++) {
		  printf("        INIT_%02X => romgen_str2bv(inst'INIT_%02X)",i,i);
		  if (i< (rom_inits - 1)) printf(",");
		  printf("\n");
		}
		printf("        )\n");
		printf("      --pragma translate_on\n");
		printf("      port map (\n");
		printf("        DO   => DATA(%d downto %d),\n",((k+1) * block_ram_width)-1,k*block_ram_width);
		if (block_ram_pwidth != 0)
		  printf("        DOP  => open,\n");
		printf("        ADDR => rom_addr,\n");
		printf("        CLK  => CLK,\n");
		printf("        DI   => \042");
		  for (i = 0; i < block_ram_width -1; i++) printf("0");
		printf("0\042,\n");
		if (block_ram_pwidth != 0) {
		  printf("        DIP  => \042");
			for (i = 0; i < block_ram_pwidth -1; i++) printf("0");
		  printf("0\042,\n");
		}
		if (format_ena == true)
		  printf("        EN   => ENA,\n");
		else
		  printf("        EN   => '1',\n");
		//
		if (format_ram16 == true)
		  printf("        SSR  => '0',\n");
		else
		  printf("        RST  => '0',\n");
		printf("        WE   => '0'\n");
		printf("        );\n");
		printf("  end generate;\n");
	  }
	} // end block ram
	//}}}

	//{{{
	if (format_array == true) {
	  if (format_clock == true)
		printf("\tprocess(clk)\n");
	  else
		printf("\tprocess(ADDR)\n");
	  printf("\tbegin\n");
	  if (format_clock == true)
		printf("\t\tif rising_edge(clk) then\n");
	  if (format_ena == true)
		printf("\t\t\tif (ena = '1') then\n  \t");
	  printf("\t\t\tdata <= ROM(to_integer(unsigned(addr)));\n");
	  if (format_ena == true)
		printf("\t\t\tend if;\n");
	  printf("\t\tend if;\n");
	  printf("\tend process;\n");
	//}}}
	} // end array

	//{{{
	if (format_case == true) {
	  if (format_clock == true)
		printf("\tprocess\n");
	  else
		printf("\tprocess(rom_addr)\n");
	  printf("\tbegin\n");
	  if (format_clock == true)
		printf("\t\twait until rising_edge(CLK);\n");
	  if (format_ena == true)
		printf("\t\tif (ena = '1') then\n");

	  printf("\t\t\tdata <= (others => '0');\n");
	  printf("\t\t\tcase rom_addr is\n");
		// stop generating RTL at the last non zero mem[] byte
		//the remaining addresses are covered by the default case statement
	  for (i = 0; i <= last_nz_addr; i ++ ) {
		printf("\t\t\t\twhen x\042%03X\042 => data <= x\042%02X\042;\n",i,mem[i]);
	  }
	  printf("\t\t\t\twhen others => data <= (others => '0');\n");
	  printf("\t\t\tend case;\n");
	  if (format_ena == true)
		printf("\t\tend if;\n");
	  printf("\tend process;\n");
	//}}}
	} // end case

	printf("end RTL;\n");

	return 0;
}
